spec/unit/actions/service_instance_delete_spec.rb

```diff
@@ -0,0 +1,109 @@
+require 'spec_helper'
+require 'actions/service_instance_delete'
+
+module VCAP::CloudController
+  RSpec.describe ServiceInstanceDelete do
+    describe '#delete' do
+      subject(:action) { described_class.new(event_repository) }
+      let(:event_repository) do
+        dbl = double(Repositories::ServiceEventRepository::WithUserActor)
+        allow(dbl).to receive(:record_user_provided_service_instance_event)
+        dbl
+      end
+
+      let!(:service_instance) do
+        si = VCAP::CloudController::UserProvidedServiceInstance.make(
+          name: 'foo',
+          credentials: {
+            foo: 'bar',
+            baz: 'qux'
+          },
+          syslog_drain_url: 'https://foo.com',
+          route_service_url: 'https://bar.com',
+          tags: %w(accounting mongodb)
+        )
+        si.label_ids = [
+          VCAP::CloudController::ServiceInstanceLabelModel.make(key_prefix: 'pre.fix', key_name: 'to_delete', value: 'value'),
+          VCAP::CloudController::ServiceInstanceLabelModel.make(key_prefix: 'pre.fix', key_name: 'tail', value: 'fluffy')
+        ]
+        si.annotation_ids = [
+          VCAP::CloudController::ServiceInstanceAnnotationModel.make(key_prefix: 'pre.fix', key_name: 'to_delete', value: 'value').id,
+          VCAP::CloudController::ServiceInstanceAnnotationModel.make(key_prefix: 'pre.fix', key_name: 'fox', value: 'bushy').id
+        ]
+        si
+      end
+
+      it 'deletes an user provided service instance from the database' do
+        subject.delete(service_instance)
+
+        expect {
+          service_instance.reload
+        }.to raise_error(Sequel::Error, 'Record not found')
+        expect(VCAP::CloudController::ServiceInstanceLabelModel.where(service_instance: service_instance)).to be_empty
+        expect(VCAP::CloudController::ServiceInstanceAnnotationModel.where(service_instance: service_instance)).to be_empty
+      end
+
+      it 'fails to delete managed service instances' do
+        instance = VCAP::CloudController::ManagedServiceInstance.make
+        expect { subject.delete(instance) }.to raise_error(ServiceInstanceDelete::NotImplementedError)
```

Comment:
Should this also check the record still exists and wasn't accidentally deleted?  Guess that's hard to make fail 🤔, though maybe it would have failed before switching on the type?

Reply:
Yeah, will do that. The intention was for this to be really short-lived, since the very next change will remove this code but we never know 1) how long that will take and 2) when CAPI might be released.